// vant/lib/notify/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _vue = require("vue");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _utils = require("../utils");

var _mountComponent = require("../utils/mount-component");

var _Notify = _interopRequireDefault(require("./Notify"));

var timer;
var instance;

function parseOptions(message) {
  return (0, _utils.isObject)(message) ? message : {
    message
  };
}

function initInstance() {
  ({
    instance
  } = (0, _mountComponent.mountComponent)({
    setup() {
      var {
        state,
        toggle
      } = (0, _mountComponent.usePopupState)();
      return () => (0, _vue.createVNode)(_Notify.default, (0, _extends2.default)({}, state, {
        'onUpdate:show': toggle
      }), null);
    }

  }));
}

function Notify(options) {
  if (!_utils.inBrowser) {
    return;
  }

  if (!instance) {
    initInstance();
  }

  options = (0, _extends2.default)({}, Notify.currentOptions, parseOptions(options));
  instance.open(options);
  clearTimeout(timer);

  if (options.duration > 0) {
    timer = setTimeout(Notify.clear, options.duration);
  }

  return instance;
}

function defaultOptions() {
  return {
    type: 'danger',
    color: undefined,
    message: '',
    onClose: undefined,
    onClick: undefined,
    onOpened: undefined,
    duration: 3000,
    className: '',
    lockScroll: false,
    background: undefined
  };
}

Notify.clear = () => {
  if (instance) {
    instance.toggle(false);
  }
};

Notify.currentOptions = defaultOptions();

Notify.setDefaultOptions = options => {
  (0, _extends2.default)(Notify.currentOptions, options);
};

Notify.resetDefaultOptions = () => {
  Notify.currentOptions = defaultOptions();
};

Notify.install = app => {
  app.use(_Notify.default);
  app.config.globalProperties.$notify = Notify;
};

Notify.Component = _Notify.default;
var _default = Notify;
exports.default = _default;
