// vant/lib/dropdown-menu/index.js
"use strict";

exports.__esModule = true;
exports.default = exports.DROPDOWN_KEY = void 0;

var _vue = require("vue");

var _utils = require("../utils");

var _use = require("@vant/use");

// Utils
// Composition
var [createComponent, bem] = (0, _utils.createNamespace)('dropdown-menu');
var DROPDOWN_KEY = 'vanDropdownMenu';
exports.DROPDOWN_KEY = DROPDOWN_KEY;

var _default = createComponent({
  props: {
    zIndex: [Number, String],
    activeColor: String,
    overlay: {
      type: Boolean,
      default: true
    },
    duration: {
      type: [Number, String],
      default: 0.2
    },
    direction: {
      type: String,
      default: 'down'
    },
    closeOnClickOutside: {
      type: Boolean,
      default: true
    },
    closeOnClickOverlay: {
      type: Boolean,
      default: true
    }
  },

  setup(props, {
    slots
  }) {
    var root = (0, _vue.ref)();
    var barRef = (0, _vue.ref)();
    var offset = (0, _vue.ref)(0);
    var {
      children,
      linkChildren
    } = (0, _use.useChildren)(DROPDOWN_KEY);
    var scrollParent = (0, _use.useScrollParent)(root);
    var opened = (0, _vue.computed)(() => children.some(item => item.state.showWrapper));
    var barStyle = (0, _vue.computed)(() => {
      if (opened.value && (0, _utils.isDef)(props.zIndex)) {
        return {
          zIndex: +props.zIndex + 1
        };
      }
    });

    var onClickAway = () => {
      if (props.closeOnClickOutside) {
        children.forEach(item => {
          item.toggle(false);
        });
      }
    };

    var updateOffset = () => {
      if (barRef.value) {
        var rect = (0, _use.useRect)(barRef);

        if (props.direction === 'down') {
          offset.value = rect.bottom;
        } else {
          offset.value = window.innerHeight - rect.top;
        }
      }
    };

    var onScroll = () => {
      if (opened.value) {
        updateOffset();
      }
    };

    var toggleItem = active => {
      children.forEach((item, index) => {
        if (index === active) {
          updateOffset();
          item.toggle();
        } else if (item.state.showPopup) {
          item.toggle(false, {
            immediate: true
          });
        }
      });
    };

    var renderTitle = (item, index) => {
      var {
        showPopup
      } = item.state;
      var {
        disabled,
        titleClass
      } = item;
      return (0, _vue.createVNode)("div", {
        "role": "button",
        "tabindex": disabled ? -1 : 0,
        "class": bem('item', {
          disabled
        }),
        "onClick": () => {
          if (!disabled) {
            toggleItem(index);
          }
        }
      }, [(0, _vue.createVNode)("span", {
        "class": [bem('title', {
          down: showPopup === (props.direction === 'down'),
          active: showPopup
        }), titleClass],
        "style": {
          color: showPopup ? props.activeColor : ''
        }
      }, [(0, _vue.createVNode)("div", {
        "class": "van-ellipsis"
      }, [item.renderTitle()])])]);
    };

    linkChildren({
      props,
      offset
    });
    (0, _use.useClickAway)(root, onClickAway);
    (0, _use.useEventListener)('scroll', onScroll, {
      target: scrollParent
    });
    return () => (0, _vue.createVNode)("div", {
      "ref": root,
      "class": bem()
    }, [(0, _vue.createVNode)("div", {
      "ref": barRef,
      "style": barStyle.value,
      "class": bem('bar', {
        opened: opened.value
      })
    }, [children.map(renderTitle)]), slots.default == null ? void 0 : slots.default()]);
  }

});

exports.default = _default;
