// vant/es/radio-group/index.js
import { createVNode as _createVNode } from "vue";
import { watch } from 'vue';
import { UnknownProp, createNamespace } from '../utils';
import { useChildren } from '@vant/use';
import { useLinkField } from '../composables/use-link-field';
var [createComponent, bem] = createNamespace('radio-group');
export var RADIO_KEY = 'vanRadio';
export default createComponent({
  props: {
    disabled: Boolean,
    iconSize: [Number, String],
    direction: String,
    modelValue: UnknownProp,
    checkedColor: String
  },
  emits: ['change', 'update:modelValue'],

  setup(props, {
    emit,
    slots
  }) {
    var {
      linkChildren
    } = useChildren(RADIO_KEY);

    var updateModelValue = value => {
      emit('update:modelValue', value);
    };

    watch(() => props.modelValue, value => {
      emit('change', value);
    });
    linkChildren({
      props,
      updateModelValue
    });
    useLinkField(() => props.modelValue);
    return () => _createVNode("div", {
      "class": bem([props.direction]),
      "role": "radiogroup"
    }, [slots.default == null ? void 0 : slots.default()]);
  }

});
