// vant/es/count-down/index.js
import { createVNode as _createVNode } from "vue";
import { watch, computed } from 'vue'; // Utils

import { createNamespace } from '../utils';
import { parseFormat } from './utils'; // Composition

import { useCountDown } from '@vant/use';
import { useExpose } from '../composables/use-expose';
var [createComponent, bem] = createNamespace('count-down');
export default createComponent({
  props: {
    millisecond: Boolean,
    time: {
      type: [Number, String],
      default: 0
    },
    format: {
      type: String,
      default: 'HH:mm:ss'
    },
    autoStart: {
      type: Boolean,
      default: true
    }
  },
  emits: ['change', 'finish'],

  setup(props, {
    emit,
    slots
  }) {
    var {
      start,
      pause,
      reset,
      current
    } = useCountDown({
      time: +props.time,
      millisecond: props.millisecond,

      onChange(current) {
        emit('change', current);
      },

      onFinish() {
        emit('finish');
      }

    });
    var timeText = computed(() => parseFormat(props.format, current.value));

    var resetTime = () => {
      reset(+props.time);

      if (props.autoStart) {
        start();
      }
    };

    watch(() => props.time, resetTime, {
      immediate: true
    });
    useExpose({
      start,
      pause,
      reset: resetTime
    });
    return () => _createVNode("div", {
      "class": bem()
    }, [slots.default ? slots.default(current.value) : timeText.value]);
  }

});
