// vant/es/row/index.js
import { createVNode as _createVNode } from "vue";
import { computed } from 'vue';
import { createNamespace } from '../utils';
import { useChildren } from '@vant/use';
var [createComponent, bem] = createNamespace('row');
export var ROW_KEY = 'vanRow';
export default createComponent({
  props: {
    align: String,
    justify: String,
    tag: {
      type: String,
      default: 'div'
    },
    gutter: {
      type: [Number, String],
      default: 0
    }
  },

  setup(props, {
    slots
  }) {
    var {
      children,
      linkChildren
    } = useChildren(ROW_KEY);
    var groups = computed(() => {
      var groups = [[]];
      var totalSpan = 0;
      children.forEach((child, index) => {
        totalSpan += Number(child.span);

        if (totalSpan > 24) {
          groups.push([index]);
          totalSpan -= 24;
        } else {
          groups[groups.length - 1].push(index);
        }
      });
      return groups;
    });
    var spaces = computed(() => {
      var gutter = Number(props.gutter);
      var spaces = [];

      if (!gutter) {
        return spaces;
      }

      groups.value.forEach(group => {
        var averagePadding = gutter * (group.length - 1) / group.length;
        group.forEach((item, index) => {
          if (index === 0) {
            spaces.push({
              right: averagePadding
            });
          } else {
            var left = gutter - spaces[item - 1].right;
            var right = averagePadding - left;
            spaces.push({
              left,
              right
            });
          }
        });
      });
      return spaces;
    });
    linkChildren({
      spaces
    });
    return () => {
      var {
        tag,
        align,
        justify
      } = props;
      return _createVNode(tag, {
        "class": bem({
          ["align-" + align]: align,
          ["justify-" + justify]: justify
        })
      }, {
        default: () => [slots.default == null ? void 0 : slots.default()]
      });
    };
  }

});
