// element-plus/es/el-element-plus/index.js
import ElAffix from '../el-affix';
export { default as ElAffix } from '../el-affix';
import ElAlert from '../el-alert';
export { default as ElAlert } from '../el-alert';
import ElAside from '../el-aside';
export { default as ElAside } from '../el-aside';
import ElAutocomplete from '../el-autocomplete';
export { default as ElAutocomplete } from '../el-autocomplete';
import ElAvatar from '../el-avatar';
export { default as ElAvatar } from '../el-avatar';
import ElBacktop from '../el-backtop';
export { default as ElBacktop } from '../el-backtop';
import ElBadge from '../el-badge';
export { default as ElBadge } from '../el-badge';
import ElBreadcrumb from '../el-breadcrumb';
export { default as ElBreadcrumb } from '../el-breadcrumb';
import ElBreadcrumbItem from '../el-breadcrumb-item';
export { default as ElBreadcrumbItem } from '../el-breadcrumb-item';
import ElButton from '../el-button';
export { default as ElButton } from '../el-button';
import ElButtonGroup from '../el-button-group';
export { default as ElButtonGroup } from '../el-button-group';
import ElCalendar from '../el-calendar';
export { default as ElCalendar } from '../el-calendar';
import ElCard from '../el-card';
export { default as ElCard } from '../el-card';
import ElCarousel from '../el-carousel';
export { default as ElCarousel } from '../el-carousel';
import ElCarouselItem from '../el-carousel-item';
export { default as ElCarouselItem } from '../el-carousel-item';
import ElCascader from '../el-cascader';
export { default as ElCascader } from '../el-cascader';
import ElCascaderPanel from '../el-cascader-panel';
export { default as ElCascaderPanel } from '../el-cascader-panel';
import ElCheckbox from '../el-checkbox';
export { default as ElCheckbox } from '../el-checkbox';
import ElCheckboxButton from '../el-checkbox-button';
export { default as ElCheckboxButton } from '../el-checkbox-button';
import ElCheckboxGroup from '../el-checkbox-group';
export { default as ElCheckboxGroup } from '../el-checkbox-group';
import ElCol from '../el-col';
export { default as ElCol } from '../el-col';
import ElCollapse from '../el-collapse';
export { default as ElCollapse } from '../el-collapse';
import ElCollapseItem from '../el-collapse-item';
export { default as ElCollapseItem } from '../el-collapse-item';
import ElCollapseTransition from '../el-collapse-transition';
export { default as ElCollapseTransition } from '../el-collapse-transition';
import ElColorPicker from '../el-color-picker';
export { default as ElColorPicker } from '../el-color-picker';
import ElContainer from '../el-container';
export { default as ElContainer } from '../el-container';
import ElDatePicker from '../el-date-picker';
export { default as ElDatePicker } from '../el-date-picker';
import ElDialog from '../el-dialog';
export { default as ElDialog } from '../el-dialog';
import ElDivider from '../el-divider';
export { default as ElDivider } from '../el-divider';
import ElDrawer from '../el-drawer';
export { default as ElDrawer } from '../el-drawer';
import ElDropdown from '../el-dropdown';
export { default as ElDropdown } from '../el-dropdown';
import ElDropdownItem from '../el-dropdown-item';
export { default as ElDropdownItem } from '../el-dropdown-item';
import ElDropdownMenu from '../el-dropdown-menu';
export { default as ElDropdownMenu } from '../el-dropdown-menu';
import ElEmpty from '../el-empty';
export { default as ElEmpty } from '../el-empty';
import ElFooter from '../el-footer';
export { default as ElFooter } from '../el-footer';
import ElForm from '../el-form';
export { default as ElForm } from '../el-form';
import ElFormItem from '../el-form-item';
export { default as ElFormItem } from '../el-form-item';
import ElHeader from '../el-header';
export { default as ElHeader } from '../el-header';
import ElIcon from '../el-icon';
export { default as ElIcon } from '../el-icon';
import ElImage from '../el-image';
export { default as ElImage } from '../el-image';
import ElImageViewer from '../el-image-viewer';
export { default as ElImageViewer } from '../el-image-viewer';
import ElInfiniteScroll from '../el-infinite-scroll';
export { default as ElInfiniteScroll } from '../el-infinite-scroll';
import ElInput from '../el-input';
export { default as ElInput } from '../el-input';
import ElInputNumber from '../el-input-number';
export { default as ElInputNumber } from '../el-input-number';
import ElLink from '../el-link';
export { default as ElLink } from '../el-link';
import ElLoading from '../el-loading';
export { default as ElLoading } from '../el-loading';
import ElMain from '../el-main';
export { default as ElMain } from '../el-main';
import ElMenu from '../el-menu';
export { default as ElMenu } from '../el-menu';
import ElMenuItem from '../el-menu-item';
export { default as ElMenuItem } from '../el-menu-item';
import ElMenuItemGroup from '../el-menu-item-group';
export { default as ElMenuItemGroup } from '../el-menu-item-group';
import ElMessage from '../el-message';
export { default as ElMessage } from '../el-message';
import ElMessageBox from '../el-message-box';
export { default as ElMessageBox } from '../el-message-box';
import ElNotification from '../el-notification';
export { default as ElNotification } from '../el-notification';
import ElOption from '../el-option';
export { default as ElOption } from '../el-option';
import ElOptionGroup from '../el-option-group';
export { default as ElOptionGroup } from '../el-option-group';
import ElPageHeader from '../el-page-header';
export { default as ElPageHeader } from '../el-page-header';
import ElPagination from '../el-pagination';
export { default as ElPagination } from '../el-pagination';
import ElPopconfirm from '../el-popconfirm';
export { default as ElPopconfirm } from '../el-popconfirm';
import ElPopover from '../el-popover';
export { default as ElPopover } from '../el-popover';
import ElPopper from '../el-popper';
export { default as ElPopper } from '../el-popper';
import ElProgress from '../el-progress';
export { default as ElProgress } from '../el-progress';
import ElRadio from '../el-radio';
export { default as ElRadio } from '../el-radio';
import ElRadioButton from '../el-radio-button';
export { default as ElRadioButton } from '../el-radio-button';
import ElRadioGroup from '../el-radio-group';
export { default as ElRadioGroup } from '../el-radio-group';
import ElRate from '../el-rate';
export { default as ElRate } from '../el-rate';
import ElRow from '../el-row';
export { default as ElRow } from '../el-row';
import ElScrollbar from '../el-scrollbar';
export { default as ElScrollbar } from '../el-scrollbar';
import ElSelect from '../el-select';
export { default as ElSelect } from '../el-select';
import ElSlider from '../el-slider';
export { default as ElSlider } from '../el-slider';
import ElStep from '../el-step';
export { default as ElStep } from '../el-step';
import ElSteps from '../el-steps';
export { default as ElSteps } from '../el-steps';
import ElSubmenu from '../el-submenu';
export { default as ElSubmenu } from '../el-submenu';
import ElSwitch from '../el-switch';
export { default as ElSwitch } from '../el-switch';
import ElTabPane from '../el-tab-pane';
export { default as ElTabPane } from '../el-tab-pane';
import ElTable from '../el-table';
export { default as ElTable } from '../el-table';
import ElTableColumn from '../el-table-column';
export { default as ElTableColumn } from '../el-table-column';
import ElTabs from '../el-tabs';
export { default as ElTabs } from '../el-tabs';
import ElTag from '../el-tag';
export { default as ElTag } from '../el-tag';
import ElTimePicker from '../el-time-picker';
export { default as ElTimePicker } from '../el-time-picker';
import ElTimeSelect from '../el-time-select';
export { default as ElTimeSelect } from '../el-time-select';
import ElTimeline from '../el-timeline';
export { default as ElTimeline } from '../el-timeline';
import ElTimelineItem from '../el-timeline-item';
export { default as ElTimelineItem } from '../el-timeline-item';
import ElTooltip from '../el-tooltip';
export { default as ElTooltip } from '../el-tooltip';
import ElTransfer from '../el-transfer';
export { default as ElTransfer } from '../el-transfer';
import ElTree from '../el-tree';
export { default as ElTree } from '../el-tree';
import ElUpload from '../el-upload';
export { default as ElUpload } from '../el-upload';
import ElVirtualList from '../el-virtual-list';
export { default as ElVirtualList } from '../el-virtual-list';
import ElSpace from '../el-space';
export { default as ElSpace } from '../el-space';
import ElSkeleton from '../el-skeleton';
export { default as ElSkeleton } from '../el-skeleton';
import ElSkeletonItem from '../el-skeleton-item';
export { default as ElSkeletonItem } from '../el-skeleton-item';
import { use, i18n } from '../locale';
import { setConfig } from '../utils/config';
import isServer from '../utils/isServer';
import dayjs from 'dayjs';

const version = '1.0.2-beta.33';

if (!isServer && !window.dayjs) {
    window.dayjs = dayjs;
}
const version$1 = version;
const locale = use;
const defaultInstallOpt = {
    size: '',
    zIndex: 2000,
};
const components = [
    ElAffix,
    ElAlert,
    ElAside,
    ElAutocomplete,
    ElAvatar,
    ElBacktop,
    ElBadge,
    ElBreadcrumb,
    ElBreadcrumbItem,
    ElButton,
    ElButtonGroup,
    ElCalendar,
    ElCard,
    ElCarousel,
    ElCarouselItem,
    ElCascader,
    ElCascaderPanel,
    ElCheckbox,
    ElCheckboxButton,
    ElCheckboxGroup,
    ElCol,
    ElCollapse,
    ElCollapseItem,
    ElCollapseTransition,
    ElColorPicker,
    ElContainer,
    ElDatePicker,
    ElDialog,
    ElDivider,
    ElDrawer,
    ElDropdown,
    ElDropdownItem,
    ElDropdownMenu,
    ElEmpty,
    ElFooter,
    ElForm,
    ElFormItem,
    ElHeader,
    ElIcon,
    ElImage,
    ElImageViewer,
    ElInput,
    ElInputNumber,
    ElLink,
    ElMain,
    ElMenu,
    ElMenuItem,
    ElMenuItemGroup,
    ElOption,
    ElOptionGroup,
    ElPageHeader,
    ElPagination,
    ElPopconfirm,
    ElPopover,
    ElPopper,
    ElProgress,
    ElRadio,
    ElRadioButton,
    ElRadioGroup,
    ElRate,
    ElRow,
    ElScrollbar,
    ElSelect,
    ElSlider,
    ElStep,
    ElSteps,
    ElSubmenu,
    ElSwitch,
    ElTabPane,
    ElTable,
    ElTableColumn,
    ElTabs,
    ElTag,
    ElTimePicker,
    ElTimeSelect,
    ElTimeline,
    ElTimelineItem,
    ElTooltip,
    ElTransfer,
    ElTree,
    ElUpload,
    ElVirtualList,
    ElSpace,
    ElSkeleton,
    ElSkeletonItem,
];
const plugins = [
    ElInfiniteScroll,
    ElLoading,
    ElMessage,
    ElMessageBox,
    ElNotification,
];
const install = (app, opt) => {
    const option = Object.assign(defaultInstallOpt, opt);
    locale(option.locale);
    if (option.i18n) {
        i18n(option.i18n);
    }
    app.config.globalProperties.$ELEMENT = option;
    setConfig(option);
    components.forEach(component => {
        app.component(component.name, component);
    });
    plugins.forEach(plugin => {
        app.use(plugin);
    });
};
var index = {
    version: version$1,
    install,
};

export default index;
export { install, locale, version$1 as version };
