// vant/lib/coupon/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _vue = require("vue");

var _utils = require("../utils");

var _constant = require("../utils/constant");

var _checkbox = _interopRequireDefault(require("../checkbox"));

var [createComponent, bem, t] = (0, _utils.createNamespace)('coupon');

function getDate(timeStamp) {
  var date = new Date(timeStamp * 1000);
  return date.getFullYear() + "." + (0, _utils.padZero)(date.getMonth() + 1) + "." + (0, _utils.padZero)(date.getDate());
}

function formatDiscount(discount) {
  return (discount / 10).toFixed(discount % 10 === 0 ? 0 : 1);
}

function formatAmount(amount) {
  return (amount / 100).toFixed(amount % 100 === 0 ? 0 : amount % 10 === 0 ? 1 : 2);
}

var _default = createComponent({
  props: {
    chosen: Boolean,
    disabled: Boolean,
    coupon: {
      type: Object,
      required: true
    },
    currency: {
      type: String,
      default: '¥'
    }
  },

  setup(props) {
    var validPeriod = (0, _vue.computed)(() => {
      var {
        startAt,
        endAt
      } = props.coupon;
      return getDate(startAt) + " - " + getDate(endAt);
    });
    var faceAmount = (0, _vue.computed)(() => {
      var {
        coupon,
        currency
      } = props;

      if (coupon.valueDesc) {
        return [coupon.valueDesc, (0, _vue.createVNode)("span", null, [coupon.unitDesc || ''])];
      }

      if (coupon.denominations) {
        var denominations = formatAmount(coupon.denominations);
        return [(0, _vue.createVNode)("span", null, {
          default: () => [currency]
        }), " " + denominations];
      }

      if (coupon.discount) {
        return t('discount', formatDiscount(coupon.discount));
      }

      return '';
    });
    var conditionMessage = (0, _vue.computed)(() => {
      var condition = formatAmount(props.coupon.originCondition || 0);
      return condition === '0' ? t('unlimited') : t('condition', condition);
    });
    return () => {
      var {
        chosen,
        coupon,
        disabled
      } = props;
      var description = disabled && coupon.reason || coupon.description;
      return (0, _vue.createVNode)("div", {
        "class": bem({
          disabled
        })
      }, [(0, _vue.createVNode)("div", {
        "class": bem('content')
      }, [(0, _vue.createVNode)("div", {
        "class": bem('head')
      }, [(0, _vue.createVNode)("h2", {
        "class": bem('amount')
      }, [faceAmount.value]), (0, _vue.createVNode)("p", {
        "class": bem('condition')
      }, [coupon.condition || conditionMessage.value])]), (0, _vue.createVNode)("div", {
        "class": bem('body')
      }, [(0, _vue.createVNode)("p", {
        "class": bem('name')
      }, [coupon.name]), (0, _vue.createVNode)("p", {
        "class": bem('valid')
      }, [validPeriod.value]), !disabled && (0, _vue.createVNode)(_checkbox.default, {
        "class": bem('corner'),
        "modelValue": chosen,
        "checkedColor": _constant.RED
      }, null)])]), description && (0, _vue.createVNode)("p", {
        "class": bem('description')
      }, {
        default: () => [description]
      })]);
    };
  }

});

exports.default = _default;
