// vant/lib/nav-bar/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _vue = require("vue");

var _utils = require("../utils");

var _constant = require("../utils/constant");

var _usePlaceholder = require("../composables/use-placeholder");

var _icon = _interopRequireDefault(require("../icon"));

// Utils
// Composition
// Components
var [createComponent, bem] = (0, _utils.createNamespace)('nav-bar');

var _default = createComponent({
  props: {
    title: String,
    fixed: Boolean,
    zIndex: [Number, String],
    leftText: String,
    rightText: String,
    leftArrow: Boolean,
    placeholder: Boolean,
    safeAreaInsetTop: Boolean,
    border: {
      type: Boolean,
      default: true
    }
  },
  emits: ['click-left', 'click-right'],

  setup(props, {
    emit,
    slots
  }) {
    var navBarRef = (0, _vue.ref)();
    var renderPlaceholder = (0, _usePlaceholder.usePlaceholder)(navBarRef, bem);

    var onClickLeft = event => {
      emit('click-left', event);
    };

    var onClickRight = event => {
      emit('click-right', event);
    };

    var renderLeft = () => {
      if (slots.left) {
        return slots.left();
      }

      return [props.leftArrow && (0, _vue.createVNode)(_icon.default, {
        "class": bem('arrow'),
        "name": "arrow-left"
      }, null), props.leftText && (0, _vue.createVNode)("span", {
        "class": bem('text')
      }, [props.leftText])];
    };

    var renderRight = () => {
      if (slots.right) {
        return slots.right();
      }

      return (0, _vue.createVNode)("span", {
        "class": bem('text')
      }, [props.rightText]);
    };

    var renderNavBar = () => {
      var {
        title,
        fixed,
        border,
        zIndex
      } = props;
      var style = {
        zIndex: zIndex !== undefined ? +zIndex : undefined
      };
      var hasLeft = props.leftArrow || props.leftText || slots.left;
      var hasRight = props.rightText || slots.right;
      return (0, _vue.createVNode)("div", {
        "ref": navBarRef,
        "style": style,
        "class": [bem({
          fixed,
          'safe-area-inset-top': props.safeAreaInsetTop
        }), {
          [_constant.BORDER_BOTTOM]: border
        }]
      }, [(0, _vue.createVNode)("div", {
        "class": bem('content')
      }, [hasLeft && (0, _vue.createVNode)("div", {
        "class": bem('left'),
        "onClick": onClickLeft
      }, [renderLeft()]), (0, _vue.createVNode)("div", {
        "class": [bem('title'), 'van-ellipsis']
      }, [slots.title ? slots.title() : title]), hasRight && (0, _vue.createVNode)("div", {
        "class": bem('right'),
        "onClick": onClickRight
      }, [renderRight()])])]);
    };

    return () => {
      if (props.fixed && props.placeholder) {
        return renderPlaceholder(renderNavBar);
      }

      return renderNavBar();
    };
  }

});

exports.default = _default;
