// ant-design-vue/lib/upload/Dragger.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _vue = require("vue");

var _propsUtil = require("../_util/props-util");

var _Upload = _interopRequireDefault(require("./Upload"));

var _interface = require("./interface");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !(0, _vue.isVNode)(s);
}

var _default2 = (0, _vue.defineComponent)({
  name: 'AUploadDragger',
  inheritAttrs: false,
  props: _interface.UploadProps,
  render: function render() {
    var _slot;

    var props = (0, _propsUtil.getOptionProps)(this);

    var height = props.height,
        restProps = __rest(props, ["height"]);

    var _a = this.$attrs,
        style = _a.style,
        restAttrs = __rest(_a, ["style"]);

    var draggerProps = _extends(_extends(_extends({}, restProps), restAttrs), {
      type: 'drag',
      style: _extends(_extends({}, style), {
        height: height
      })
    });

    return (0, _vue.createVNode)(_Upload.default, draggerProps, _isSlot(_slot = (0, _propsUtil.getSlot)(this)) ? _slot : {
      default: function _default() {
        return [_slot];
      }
    });
  }
});

exports.default = _default2;
