// vant/es/calendar/components/Day.js
import { createVNode as _createVNode } from "vue";
import { computed } from 'vue';
import { createNamespace } from '../../utils';
import { bem } from '../utils';
var [createComponent] = createNamespace('calendar-day');
export default createComponent({
  props: {
    color: String,
    index: Number,
    rowHeight: String,
    offset: {
      type: Number,
      default: 0
    },
    item: {
      type: Object,
      required: true
    }
  },
  emits: ['click'],

  setup(props, {
    emit
  }) {
    var style = computed(() => {
      var {
        item,
        index,
        color,
        offset,
        rowHeight
      } = props;
      var style = {
        height: rowHeight
      };

      if (item.type === 'placeholder') {
        style.width = '100%';
        return style;
      }

      if (index === 0) {
        style.marginLeft = 100 * offset / 7 + "%";
      }

      if (color) {
        switch (item.type) {
          case 'end':
          case 'start':
          case 'start-end':
          case 'multiple-middle':
          case 'multiple-selected':
            style.background = color;
            break;

          case 'middle':
            style.color = color;
            break;
        }
      }

      return style;
    });

    var onClick = () => {
      if (props.item.type !== 'disabled') {
        emit('click', props.item);
      }
    };

    var renderContent = () => {
      var {
        item,
        color,
        rowHeight
      } = props;
      var {
        type,
        text,
        topInfo,
        bottomInfo
      } = item;

      var TopInfo = topInfo && _createVNode("div", {
        "class": bem('top-info')
      }, {
        default: () => [topInfo]
      });

      var BottomInfo = bottomInfo && _createVNode("div", {
        "class": bem('bottom-info')
      }, {
        default: () => [bottomInfo]
      });

      var Nodes = [TopInfo, text, BottomInfo];

      if (type === 'selected') {
        return _createVNode("div", {
          "class": bem('selected-day'),
          "style": {
            width: rowHeight,
            height: rowHeight,
            background: color
          }
        }, {
          default: () => [Nodes]
        });
      }

      return Nodes;
    };

    return () => {
      var {
        type,
        className
      } = props.item;
      return _createVNode("div", {
        "role": "gridcell",
        "style": style.value,
        "class": [bem('day', type), className],
        "tabindex": type === 'disabled' ? undefined : -1,
        "onClick": onClick
      }, [renderContent()]);
    };
  }

});
