// @vant/use/dist/esm/useRelation/useChildren.js
import _extends from "@babel/runtime/helpers/esm/extends";
import { isVNode, provide, reactive, getCurrentInstance } from 'vue';
export function flattenVNodes(children) {
  var result = [];

  var traverse = function traverse(children) {
    if (Array.isArray(children)) {
      children.forEach(function (child) {
        if (isVNode(child)) {
          var _child$component;

          result.push(child);

          if ((_child$component = child.component) != null && _child$component.subTree) {
            traverse(child.component.subTree.children);
          }

          if (child.children) {
            traverse(child.children);
          }
        }
      });
    }
  };

  traverse(children);
  return result;
} // sort children instances by vnodes order

export function sortChildren(parent, publicChildren, internalChildren) {
  var vnodes = flattenVNodes(parent.subTree.children);
  internalChildren.sort(function (a, b) {
    return vnodes.indexOf(a.vnode) - vnodes.indexOf(b.vnode);
  });
  var orderedPublicChildren = internalChildren.map(function (item) {
    return item.proxy;
  });
  publicChildren.sort(function (a, b) {
    var indexA = orderedPublicChildren.indexOf(a);
    var indexB = orderedPublicChildren.indexOf(b);
    return indexA - indexB;
  });
}
export function useChildren(key) {
  var publicChildren = reactive([]);
  var internalChildren = reactive([]);
  var parent = getCurrentInstance();

  var linkChildren = function linkChildren(value) {
    var link = function link(child) {
      if (child.proxy) {
        internalChildren.push(child);
        publicChildren.push(child.proxy);
        sortChildren(parent, publicChildren, internalChildren);
      }
    };

    var unlink = function unlink(child) {
      var index = internalChildren.indexOf(child);
      publicChildren.splice(index, 1);
      internalChildren.splice(index, 1);
    };

    provide(key, _extends({
      link: link,
      unlink: unlink,
      children: publicChildren,
      internalChildren: internalChildren
    }, value));
  };

  return {
    children: publicChildren,
    linkChildren: linkChildren
  };
}
