// vant/es/coupon/index.js
import { createVNode as _createVNode } from "vue";
import { computed } from 'vue';
import { padZero, createNamespace } from '../utils';
import { RED } from '../utils/constant';
import Checkbox from '../checkbox';
var [createComponent, bem, t] = createNamespace('coupon');

function getDate(timeStamp) {
  var date = new Date(timeStamp * 1000);
  return date.getFullYear() + "." + padZero(date.getMonth() + 1) + "." + padZero(date.getDate());
}

function formatDiscount(discount) {
  return (discount / 10).toFixed(discount % 10 === 0 ? 0 : 1);
}

function formatAmount(amount) {
  return (amount / 100).toFixed(amount % 100 === 0 ? 0 : amount % 10 === 0 ? 1 : 2);
}

export default createComponent({
  props: {
    chosen: Boolean,
    disabled: Boolean,
    coupon: {
      type: Object,
      required: true
    },
    currency: {
      type: String,
      default: '¥'
    }
  },

  setup(props) {
    var validPeriod = computed(() => {
      var {
        startAt,
        endAt
      } = props.coupon;
      return getDate(startAt) + " - " + getDate(endAt);
    });
    var faceAmount = computed(() => {
      var {
        coupon,
        currency
      } = props;

      if (coupon.valueDesc) {
        return [coupon.valueDesc, _createVNode("span", null, [coupon.unitDesc || ''])];
      }

      if (coupon.denominations) {
        var denominations = formatAmount(coupon.denominations);
        return [_createVNode("span", null, {
          default: () => [currency]
        }), " " + denominations];
      }

      if (coupon.discount) {
        return t('discount', formatDiscount(coupon.discount));
      }

      return '';
    });
    var conditionMessage = computed(() => {
      var condition = formatAmount(props.coupon.originCondition || 0);
      return condition === '0' ? t('unlimited') : t('condition', condition);
    });
    return () => {
      var {
        chosen,
        coupon,
        disabled
      } = props;
      var description = disabled && coupon.reason || coupon.description;
      return _createVNode("div", {
        "class": bem({
          disabled
        })
      }, [_createVNode("div", {
        "class": bem('content')
      }, [_createVNode("div", {
        "class": bem('head')
      }, [_createVNode("h2", {
        "class": bem('amount')
      }, [faceAmount.value]), _createVNode("p", {
        "class": bem('condition')
      }, [coupon.condition || conditionMessage.value])]), _createVNode("div", {
        "class": bem('body')
      }, [_createVNode("p", {
        "class": bem('name')
      }, [coupon.name]), _createVNode("p", {
        "class": bem('valid')
      }, [validPeriod.value]), !disabled && _createVNode(Checkbox, {
        "class": bem('corner'),
        "modelValue": chosen,
        "checkedColor": RED
      }, null)])]), description && _createVNode("p", {
        "class": bem('description')
      }, {
        default: () => [description]
      })]);
    };
  }

});
